src/dai.rs

```diff
@@ -1,20 +1,43 @@
 use crate::publish::NominalMulf64;
+use num::{BigUint, FromPrimitive, ToPrimitive};
+use std::ops::Mul;
+use std::ops::Div;
 
-#[derive(Debug, Copy, Clone, Ord, PartialOrd, PartialEq, Eq)]
-pub struct Amount(u64);
+pub const DAI_DEC: u64 = 1_000_000_000_000_000_000;
+
+#[derive(Debug, Clone, Ord, PartialOrd, PartialEq, Eq)]
+pub struct Amount(BigUint);
+
+impl Amount {
+
+    #[allow(clippy::cast_precision_loss)] // It's ok because it just means we are applying slightly more than the given spread
+    pub fn from_dai(dai: f64) -> Self {
+        let dai_dec = BigUint::from(DAI_DEC);
+        let precision = 10u64^14;
+        let value = (dai * precision as f64).round();
+        let big_u_int = BigUint::from_f64(value).unwrap();
+
+        Amount(dai_dec.mul(big_u_int).div(precision))
+    }
```

Comment:
If these functions are meant to be used in production code, I would suggest testing them with `proptest`: https://docs.rs/proptest/0.10.0/proptest/

Reply:
To elaborate, I am referring to doing round-trip tests and/or test that simply check that you actually can't hit those `unwrap`s.